### ISSUE TITLE
Catkin needs better debugging tools or it should work in "isolated" mode by default

### ISSUE THREAD

**Issue by jbohren** [ros/catkin#539]
I keep running into issues from crossover side-effects when using catkin_make which do not occur when using catkin_make_isolated.

I don't think the speedup gained from catkin_make is worth it since it takes so long to debug errors when they occur. Either we need better ways to debug errors, or we need make catkin_make_isolated the default behavior.

The amount of time I have wasted trying to debug catkin builds is absurd.

@tfoote @tkruse

**Comment by tkruse** [comment]
If CMI were the default, I believe maintainers (where those even exist anymore) would stop checking their CMakeLists for compatibility with catkin_make, which could break the buildfarm if that still runs using catkin_make. This would also break other users' workspaces. So this cannot just be about making one the "default", but it has to be about getting rid of catkin_make (or getting rid of non-isolation) if usability issues are significant enough. For c_m_i this would also mean finally implementing rosbuild-like parallelity (Because build speed was a main catkin design goal, and without parallelity cmi would waste too much time).

OSRF has to decide whether non-isolation is a design mistake or not, and admit if that is so. Since admitting mistakes quickly becomes a personal and political issue rather than a technical one, this whole subject must be dealt with more sensitivity than usual.

And since volunteers to take up ROS package maintainership or orphaned packages are not exactly queueing up, catkin usability is a serious issue.

It would help if problems caused by lack of isolation were not just fixed, but listed on a wiki site (e.g. a list of links to github issues), so it is visible what is hurting the community most, and provide better error reporting if possible. That would be a better basis for dicsussion than personal experiences and opinions.
Also this could be refined to help people learn detect flaws in CMakeLists.txt files easier.

I had tried to predict sources of errors before catkin was delivered in hydro on the catkin review page (http://wiki.ros.org/catkin/Reviews/2012-08-01_API_Review#On_Catkin_cmake_conflicts), but that was a theoretical exercise, I do not know what causes most pain in practice.

Some links to previous discussions on isolation:

https://groups.google.com/d/msg/ros-sig-buildsystem/4u3hQ2aCWD0/ntfIit5JpxQJ

https://groups.google.com/d/msg/ros-sig-buildsystem/xCYO2EOgd0M/14RBk-n8eGEJ

http://wiki.ros.org/catkin/Reviews/2012-08-01_API_Review

Some background on cmi (catkin_make_isolated):
It was defined in REP 134: https://github.com/ros-infrastructure/rep/blob/master/rep-0134.rst
Short version: cmi was introduced mostly to allow having catkin workspaces mixing standard and non-standard catkin packages in such a way that a standard catkin package can depend on a non-standard catkin package (In this situation catkin_make currently refuses to build).

**Comment by dirk-thomas** [comment]
@jbohren Can you please provide more information on what issues from crossover side-effects you have run into?

**Comment by wjwwood** [comment]
@jbohren I may be out of the loop on this, but with no context I find it difficult to understand what specific issue(s) you are having or how I could possible address them.

You propose we have better debugging tools, do you have specific ideas for such tools? "Making better debug tools" is not something I can act on and have any chance of meeting you needs.

You also suggested that we make `catkin_make_isolated` the "default". I'm not sure I understand what you mean by "default" behavior, you mean what we recommend users use or replacing/renaming the tools?

How would you propose we implement this, remove `catkin_make` as we know it and rename `catkin_make_isolated` to `catkin_make` or some other migration?

What about the scenario where users would like to work outside of our build helper scripts (i.e. `catkin_make_isolated` and `catkin_make`) and invoke cmake directly on a workspace or single package?

**Comment by jbohren** [comment]
> @jbohren I may be out of the loop on this, but with no context I find it difficult to understand what specific issue(s) you are having or how I could possible address them.
> 
> You propose we have better debugging tools, do you have specific ideas for such tools? "Making better debug tools" is not something I can act on and have any chance of meeting you needs.

So one of the most common problems is one package changing the state of the system (changing env and cmake cache variables) which another package doesn't expect to change. Is this something that could be debugged in the configuration stage? Could catkin inject CMake code in-between each inclusion of CMake code to "watch" a given env/cache variable or set of variables? 

A big part of this is also the proposed "catkin lint" ( #153 , https://www.bountysource.com/issues/741065-catkin-lint ) which could also detect problems even _before_ configuration time.

Part of the frustration is catkin necessitates a need for new debugging patterns. Approaches like switching back-and-forth between catkin_make and catkin_make_isolated to figure out which packages are causing problems have proven useful, but also painful.

> You also suggested that we make catkin_make_isolated the "default". I'm not sure I understand what you mean by "default" behavior, you mean what we recommend users use or replacing/renaming the tools?

Yeah, I think we should either provide people the tools they need to more effectively debug the failure modes of catkin, or that we shouldn't recommend that they use an approach which can end up costing days of debugging in complex workspaces.

> How would you propose we implement this, remove catkin_make as we know it and rename catkin_make_isolated to catkin_make or some other migration?

I think that sort of discussion could come later, if necessary, I'm just worried about people trying out ROS, getting frustrated, and giving up.

> What about the scenario where users would like to work outside of our build helper scripts (i.e. catkin_make_isolated and catkin_make) and invoke cmake directly on a workspace or single package?

Then they're unlikely to run into the sort of problems I'm talking about. (:

**Comment by tkruse** [comment]
On Sat, Oct 19, 2013 at 1:30 AM, William Woodall
notifications@github.comwrote:

> @jbohren https://github.com/jbohren I may be out of the loop on this,
> but with no context I find it difficult to understand what specific
> issue(s) you are having or how I could possible address them.
> 
> @jbohren, it might help to point to commits /pull requests where you have
> recently fixed build problems after spending much time finding them.
> 
> What about the scenario where users would like to work outside of our
> build helper scripts (i.e. catkin_make_isolated and catkin_make) and
> invoke cmake directly on a workspace or single package?
> 
> Do you mean to say that ROS needs to have a non-isolated approach (central
> CMakeLists.txt in catkin_ws/src) to support the use-case of users wanting
> to build a large set of individual packages using only cmake invocations
> just as for a single package?

Invoking the build for individual packages seems important to several users
(see questions about this on answers.ros.org), and being able to bypass
catkin is a bonus in a certain way, since this allows people to eliminate
catkin code as a source of errors. However I don't see how that would be of
any help for a whole workspace.

**Comment by jbohren** [comment]
> > On Sat, Oct 19, 2013 at 1:30 AM, William Woodall notifications@github.comwrote:
> >  @jbohren https://github.com/jbohren I may be out of the loop on this,
> >  but with no context I find it difficult to understand what specific
> >  issue(s) you are having or how I could possible address them.
> 
> @jbohren, it might help to point to commits /pull requests where you have
> recently fixed build problems after spending much time finding them.

One such example was where the KDL CMake configuration file was re-setting $PKG_CONFIG_PATH whenever it was configured to the place where KDL was installed. This means that pkg-config wasn't able to find any packages which weren't installed to /opt/ros/hydro. It turns out @vrabaud and I both encountered this one at the same time and fixed it in parallel. 

There were a lot of challenges that came with making it so Orocos/RTT code could be built in a catkin workspace. I originally expected that Catkin would ease integration with 3rd-party toolchains than rosbuild, but the inherent complexity in Catkin which enables it to be built in so many different ways seemed to make it a lot harder. Tracking down which dependencies are needed by which targets across packages is also something that can be hard to debug if someone else made a mistake when writing their packages, assuming that build products would be available. This has the effect of code building on one computer in one order, but not building on another computer in a different order (which can happen if the workspace changes slightly).

Both of these, and other problems are resolved by using `catkin_make_isolated` since it has neither the cross-talk issues nor the cross-package dependency requirements for proper build ordering.

**Comment by jbohren** [comment]
> One such example was where the KDL CMake configuration file was re-setting $PKG_CONFIG_PATH whenever it was configured to the place where KDL was installed. This means that pkg-config wasn't able to find any packages which weren't installed to /opt/ros/hydro. It turns out @vrabaud and I both encountered this one at the same time and fixed it in parallel.

More specifically, the bug in the KDL cmake config (fixed here: https://github.com/orocos/orocos_kinematics_dynamics/commit/e1bf670dc6fe878efea5b78e99e4d76327610749) would re-set `$PKG_CONFIG_PATH` simply when the user _depended_ on `orocos_kdl`. This means that while the observed problem was that `$PKG_CONFIG_PATH` was being re-set, there was _nothing_ in my workspace of about 150 packages which mentioned that variable. it took three days to track down that the kdl config file was the problem
in this workspace with about 150 packages and it involved going through every package and outputting `$PKG_CONFIG_PATH` starting with the package which was failing (which was any package that relied on pkg-config. 

Normally, this change would only affect a single package's single invocation of CMake, so the search space for the problem would be limited to a single CMakeLists.txt file. Instead, catkin created the possibility that any of the 150 CMake files in the workspace and their included files could be changing the environment variables. The bug didn't appear in `catkin_make_isolated` because it didn't cause any problems for the particular packages using KDL, but it caused problems in the packages that had nothing to do with KDL at all.

This was further complicated because I had two workspaces: an isolated workspace and a "normal" workspace, with the latter building against the former.

**Comment by tkruse** [comment]
So as another way to put it, the lack of isolation in workspaces does not just make the given CmakeLists.txt of packages interact (that which a catkinlint tool could maybe catch, maybe not), but also the cmake files provided by 3rd party libraries. An since those 3rd-party provided libraries are usually used at the top of cmake files, bugs due to bad practice more easily surface in catkin workspaces due to lack of isolation than in isolated cmake projects.
And finding which 3rd party library is responsible if the symptoms appear in a given catkin workspace with many packages is a huge effort. 

It would be nice also to have a concrete example for the task dependency topic. This thread once discussed the general problematics of cross-package cmake task dependencies: https://groups.google.com/d/msg/ros-sig-buildsystem/dvVO5QCHBLM/QtXUOf7EOhQJ
Not sure whether the issues you faced were related to cmake's "add_dependencies".

**Comment by dirk-thomas** [comment]
@jbohren I do understand your frustration when having to deal with such hard to debug problems. I can see why catkin_make_isolated is easier for users to use and may be it would be a good idea to recommend users to use that tool instead of catkin_make.

On the other hand I do see a lot of advantages to the single CMake invocation. It is not only faster to configure and build but also integrates better which external tools like IDEs. E.g. it is very convenient to be able to go to the build folder and just invoke any of the targets explicitly.

From the problems you have mentioned I can sadly not imagine a "solution" within catkin to address any of them (besides using cmi instead of cm). Most of them come simply from the intrinsic complexity of the task. Some others are related to bad written CMake - which is really easy to achieve since CMake is not a greatly structured language - more like c - powerful but dangerous. If we can come up with any feasible changes I am happy to implement them to make the system better / easier to use.

**Comment by vrabaud** [comment]
Same, I agree with @dirk-thomas  but understand @jbohren  . I found the PKG_CONFIG_PATH error fairly quickly ( :p ) but I agree it was a hard one.

Another bug I had the other day was because find_package(catkin) was called after find_package(OpenCV) thus setting GTEST_FOUND too late and adding some tests to OpenCV_LIBRARIES.

Somebody mentioned earlier a debug mode that would record variables changed/used by catkin (or displayed at the beginning/end of catkin) in the find_package(catkin) and though it will never catch everything, it would be a great start.

Even if that is not code, maybe just a document listing the variables changed/used by catkin would be useful (those are the only things that can really create a CMake problem). There are fairly few in the end and the two problems I ever had between catkin and 3rd party packages were due to some of them (PKG_CONFIG_PATH and GTEST_FOUND).

(maybe such a document exists, I have not RTFMed catkin in a while :) )

**Comment by jbohren** [comment]
@dirk-thomas 

> On the other hand I do see a lot of advantages to the single CMake invocation. It is not only faster to configure and build but also integrates better which external tools like IDEs. E.g. it is very convenient to be able to go to the build folder and just invoke any of the targets explicitly.
1. Being "faster to configure and build" is worthless if it creates the possibility an un-knowable amount of delay in the development cycle.
2. With respect to the convenience of going "to the build folder and just invoke any of the targets explicitly," I don't see how this necessitates a single call to CMake since you can do the same exact thing with catkin_make_isolated.

I understand that there are potential benefits from the design of catkin_make, but are they _actual_ benefits when incorporated into the real development workflow? Who was this system designed for exactly? Who are the intended users of Catkin? The requirements that Catkin make it easier to use IDEs with ROS was not one of the use cases in the only list of use cases I know of (http://wiki.ros.org/catkin/Reviews/2012-08-01_API_Review?highlight=%28PackageReviewCategory%29#Use_cases).

@dirk-thomas 

> From the problems you have mentioned I can sadly not imagine a "solution" within catkin to address any of them (besides using cmi instead of cm). 

This is exactly why I included that proposition (to advise people to use cmi instead of cm) in the title of this ticket. I think what might just happen is people will find that cmi works and cm doesn't, and then you will end up with so many packages in the community that don't work with cm that you will either be forced to deprecate it or people will just stop using ROS (and Catkin) entirely.

@dirk-thomas 

> Some others are related to bad written CMake - which is really easy to achieve since CMake is not a greatly structured language - more like c - powerful but dangerous.

This might be central to the problem. Catkin has now made proficiency in CMake a pre-requisite to building software with ROS. When I started to use ROS in 2009, I had never used CMake before, but that didn't stop me from being able to develop with ROS. 

At the end of the day, I feel like every change to the core ROS systems should have to answer the question:

**_Does this change make it easier to make robotics happen?**_

I still don't think the current version of Catkin does this, and I think OSRF should be seriously concerned about that.

**Comment by dirk-thomas** [comment]
1. It _does_ work very well for a significant part of the released ROS packages in Hydro. And if it does cm has benefits compared to cmi.
2. Calling a specific make target is something you exactly can not do with cmi. Since there are multiple build folders you have to go to a specific one first - which you either don't want to care about or just don't know which is the corresponding package.

And depending on what kind of workspace you are working on the performance factor also matters.

May be you should propose this on the buildsystem SIG. if the majority is in favor of recommending cmi over cm or even only supporting cmi that is totally fine for me.

**Comment by wjwwood** [comment]
> This might be central to the problem. Catkin has now made proficiency in CMake a pre-requisite to building software with ROS. When I started to use ROS in 2009, I had never used CMake before, but that didn't stop me from being able to develop with ROS.

I agree that we need to work on the usability of catkin, but requiring no proficiency in CMake should not a high priority in my opinion. If you are not going to release or install your code then you can skate around a lot of the details and use `rosbuild` or `catkin_simple`, both of which require significantly less CMake knowledge.

By the way, We recently spent some time deploying `catkin` into a non-ROS system using `catkin_simple`. It worked quite well, but it still needs some polish before it is ready for community consumption.

Either way, for your problems stated above using something like `rosbuild` or `catkin_simple` would not have helped. You only options, as I see it, are to make the package you are talking about "plain cmake" which forces you to use CMI always, or deal with the issues of building a previously standalone CMake project as part of a larger system. You would run into exactly the same issues when integrating this previously CMake project into a larger CMake project.

As @vrabaud said, I think better documentation about which variables should not be set within a catkin package, which ones are used, and what they do could help with this. Maybe we could devise some kind of tool which tried to detect bad behavior from a catkin package which using cm.

> Being "faster to configure and build" is worthless if it creates the possibility an un-knowable amount of delay in the development cycle.

I'm not sure I agree with this, it's like saying "release mode for C++ is useless because you might run into an issue and need to debug it." In this case you would switch to a different mode which makes debugging easier. I see CMI in the same light. Maybe cmi should be the preference, but eliminating cm is a bad idea in my opinion.

> I understand that there are potential benefits from the design of catkin_make, but are they actual benefits when incorporated into the real development workflow?

I use it daily, @dirk-thomas has built all of the released catkin packages in hydro in one invocation before, just to try and find issues like this. Even though I wrote cmi originally, I personally prefer using cm over cmi if I can manage it, going so far as to make multiple workspaces to split up the packages. I find it faster and more conventional.

> Who was this system designed for exactly?

For people releasing and installing their packages, but we would like it to meet everyone's needs, and though the process towards that is slow, we are not obstructing it needlessly, we just want to make sure it is done correctly.

> The requirements that Catkin make it easier to use IDEs with ROS was not one of the use cases in the only list of use cases I know of (http://wiki.ros.org/catkin/Reviews/2012-08-01_API_Review?highlight=%28PackageReviewCategory%29#Use_cases).

That does not make it invalid. Lots of users, mostly newer or less experienced uses, use IDE's like eclispe. I can also tell you that integrating auto-completion for C++ like `SublimeClang` into my editor is much easier on a catkin_make workspace.

**Comment by jbohren** [comment]
> By the way, We recently spent some time deploying catkin into a non-ROS system using catkin_simple. It worked quite well, but it still needs some polish before it is ready for community consumption.

Yeah, I saw this, and I'm really excited that it's getting more attention!

> Maybe we could devise some kind of tool which tried to detect bad behavior from a catkin package which using cm.

Right, this could be considered part of the goal of #153, would it not?

> In this case you would switch to a different mode which makes debugging easier. I see CMI in the same light. Maybe cmi should be the preference, but eliminating cm is a bad idea in my opinion.

I never said to eliminate it, and in my experience, cmi is also what I use when I need to debug cm. One of the problems with the current workflow / tutorials / troubleshooting guides is that that pattern isn't documented. This is why I say "catkin needs better debugging tools."

> > I understand that there are potential benefits from the design of catkin_make, but are they actual benefits when incorporated into the real development workflow?
> > I use it daily, @dirk-thomas has built all of the released catkin packages in hydro in one invocation before, just to try and find issues like this. Even though I wrote cmi originally, I personally prefer using cm over cmi if I can manage it, going so far as to make multiple workspaces to split up the packages. I find it faster and more conventional.

I agree, cm, wacky as it may be, is really cool when it works, and it's really fast. When I say "real development workflow" I mean researchers/engineers working in a lab who didn't themselves write catkin and who are trying to piece together facets of code of various levels of quality from numerous institutions. 

---

I'm sorry if we got off topic, but I didn't create this ticket because I think the catkin developers are unaware of the failure modes of catkin nor did I create this because I think the catkin developers don't care. I created this because I want to figure out how we can get a handle on the failure modes by creating catkin debugging modes, tools and patterns.

**Comment by tkruse** [comment]
Hi William,

On Tue, Oct 22, 2013 at 2:40 AM, William Woodall
notifications@github.comwrote:

> Either way, for your problems stated above using something like rosbuildor
> catkin_simple would not have helped. You only options, as I see it, are
> to make the package you are talking about "plain cmake" which forces you to
> use CMI always, or deal with the issues of building a previously standalone
> CMake project as part of a larger system.

As I tried to clarify, @jbohren did not have the problem of building a
catkinized KDL. He installed KDL independently outside his catkin
workspace. So turning KDL into a plain cmake package for catkin is an
option but irrelevant to this discussion.
Jon tried to build one or many catkin packages that used KDL, but turning
those into plain cmake was also not an option, because he could not know
which of his 150 packages had either an invalid CMakeLists.txt themselves,
or imported an invalid one, or whether some other catkin bug caused his
build failure.

An option would have been for him to just stop using catkin_make, and leave
it to the rest of the world to deal with this build problem.

As @vrabaud https://github.com/vrabaud said, I think better documentation

> about which variables should not be set within a catkin package, which ones
> are used, and what they do could help with this. Maybe we could devise some
> kind of tool which tried to detect bad behavior from a catkin package which
> using cm.
> 
> This might be a detail, but again, Jon ran into a problem with a cmake file
> that was not part of a catkin package. Users can of course also run into
> similar issues with catkin packages, but the larger problem here is the
> wealth of 3rd party non-catkin libraries which may each have defects in
> their provided cmake files which only surface when used in a catkin
> workspace. That may be any other robotics middleware to integrate with,
> drivers, simulators, and so on.
> 
>  Being "faster to configure and build" is worthless if it creates the
> possibility an un-knowable amount of delay in the development cycle.
> 
> I'm not sure I agree with this, it's like saying "release mode for C++ is
> useless because you might run into an issue and need to debug it." In this
> case you would switch to a different mode which makes debugging easier. I
> see CMI in the same light. Maybe cmi should be the preference, but
> eliminating cm is a bad idea in my opinion.
> 
> If cmi is the default, users will stop caring about proper catkin rules for
> not using global variables and being careful about names. Basically by
> making cmi the default, many packages will effectively turn into "plain
> cmake" packages, in that they will violate catkin specific rules and only
> build with cmi. And assigning bugs is very complex for most people, with
> 150 packages inyour workspace, and a build failure in any of them, you have
> little way of knowing which of the 150 packages did something it should not
> have done, so you cannot easily create a ticket on a package. And to
> reproduce the build failure, you may have to recreate the whole workspace
> of 150 packages, until the culprit package was identified.

cm would then only be a valuable tool for people who want to develop a
clean subset of all catkin packages in the world. That may still be a valid
usecase, but it has to be understood as such.

 I understand that there are potential benefits from the design of

> catkin_make, but are they actual benefits when incorporated into the real
> development workflow?
> 
> I use it daily, @dirk-thomas https://github.com/dirk-thomas has built
> all of the released catkin packages in hydro in one invocation before, just
> to try and find issues like this. Even though I wrote cmi originally, I
> personally prefer using cm over cmi if I can manage it, going so far as to
> make multiple workspaces to split up the packages. I find it faster and
> more conventional.
> 
> Of course cm is much faster than cmi, given that cmi does not even have
> rosbuild-like parallelity.  Why compare speed until that is changed?
> And Jon just expressed he lost an "absurd amount of time" fixing problems
> caused by lack of isolation, so saying "cmi is faster" does not seem to
> imply that "cmi helps ROS developers save time".

There are also theoretically other solutions, cmi does not necessarily have
to build every package in isolation. A concept of "build group" or
"compatible group" could be defined for cmi to decide which packages may
share the same build. Or the subfolder structure of catkin workspaces could
be used by putting subsets of catkin packages that are fine with lack of
isolation into the same subfolder with a common top-level CMakeLists.txt.
Something like "controlled lack of isolation", siliar to workspace
chaining, but usable for the common ROS developer.

>  The requirements that Catkin make it easier to use IDEs with ROS was not
> one of the use cases in the only list of use cases I know of (
> http://wiki.ros.org/catkin/Reviews/2012-08-01_API_Review?highlight=%28PackageReviewCategory%29#Use_cases
> ).
> 
> That does not make it invalid. Lots of users, mostly newer or less
> experienced uses, use IDE's like eclispe. I can also tell you that
> integrating auto-completion for C++ like SublimeClang into my editor is
> much easier on a catkin_make workspace.
> 
> The requirements I collected there were mostly guesses by me, written after
> the fact, based on code commens, discussions, etc. catkin was not build
> based on written requirement specs (like most software), at least I have
> never found any. Good IDE integration is quite important, and I understand
> the "single cmake project" nature of default catkin workspaces can make it
> a breeze to work with an IDE. So we can add this requirement to the review
> wiki page.
> Not sure how much effort it is else to set up an iDE with many catkin
> packages such that code completion and traversal works. Also not sure
> whether it is possible to trick the IDE in believing it builds a single
> CMake project, but then using cmi instead to build.

**Comment by jbohren** [comment]
@tkruse 

> An option would have been for him to just stop using catkin_make, and leave
> it to the rest of the world to deal with this build problem.

Yeah, it's worth mentioning that when I get stuck debugging catkin-related issues, nearly everyone around me tells me that none of this is worth it, and I should just give up.

> If cmi is the default, users will stop caring about proper catkin rules for
> not using global variables and being careful about names. Basically by
> making cmi the default, many packages will effectively turn into "plain
> cmake" packages, in that they will violate catkin specific rules and only
> build with cmi.

This is a good point. On the ROS wiki page for Catkin, it says that "catkin was designed to be more conventional than rosbuild." However, Catkin imposes numerous constraints on buildscript design which the buildscript writer for an isolated CMake project never seems to have to worry about. This might be a good thing to do for a tightly integrated system like Microsoft Robotics Studio or ERSP, but ROS is not such a system. I don't think researchers using ROS should have to worry about conflicting target names, library names, or executable names until they want OSRF to host their debian packages. This is why we have tools like `rosrun` and `roslaunch`, to scope such potentially-conflicting names.

**Comment by tfoote** [comment]
Hi everyone, 

Watching this discussion it looks to me to have lost its focus on making improvements to the system. 

From the top of this post I can see one clear takeaway which is that it is sometimes hard to debug issues in a catkin workspace due to side effects in other packages.  Unfortunately a lot of these side effects are more a problem of poor CMake usage rather than anything catkin related.  (The one concrete example in this thread is KDL overrideing $PKG_CONFIG_PATH. This is clearly a bug in the implementation which was previously not exposed, and was patched promptly.) We chose CMake as it was evaluated as the best buildsystem to meet our needs and is at this point one of the most popular. (Note this does not mean that we don't have major complaints about it, just that we have the fewest complaints about it compared to the features it offers.) Unfortunately CMake is designed in a way which promotes the use of global variables and has also developed a very lax culture around style, cleanlyness, and standards for it's users. Without these fundimentals bringing together any two existung CMake projects can be a pain as they can collide in hard to debug ways.  

That being said, catkin_make is a very powerful tool.  It was designed to provide a lot of the capabilities which developers did not want to loose from the rosbuild environment but exposes the full power of CMake across the whole workspace. This enables individual target and file level dependency tracking across the whole workspace, making incremental rebuilds very quick across multiple packages. This means that users can trigger full "rebuilds" and know that their binaries are all up to date, without worrying about how long it will take. (For example in the past many people would manually move to certain packages and call make instead of calling rosmake since they didn't want to wait for the overhead of touching every package. The same issue happens if you call catkin_make_isolated, since it uses the isolated workspaces it must recrawl each workspace sequentially.) 

The power of catkin comes from leveraging CMake. However with that comes the unstructuredness of CMake. As the complexity and unintentional side effecting is easy to do in CMake we should definitely work on making these things easier to debug. There are users who do not need the full power of the tool, and for them running in small protected sandboxes can be useful, but requiring that all users run in the sandboxes is not a great solution. Nodelets follow a similar story.  All nodelets can be run alone, but they don't get the benefits of being in the same process, and to enable the nodelets to run in the same process they have to follow certain conventions. If you don't follow those conventions you have a node which cannot be run in the same process. This would be like saying that nobody can use nodelets because a memory error in one nodelet can effect the other nodelets. 

There are users for whom the full power is not as valuable, such as new students using ROS. But new students using ROS I do not expect to be integrating 3rdparty CMake based libraries which have bugs causing breaking side effects. 

I'd like to close this issue, it's very broad and unactionable. From the discussion I've filed a new ticket: https://github.com/ros/catkin/issues/543 which I think can help with the debugging of catkin workspaces based on the concrete examples presented here. 

In the future, lets focus on specific issues where things can be improved. Whether it being a debugging a specific type of issue is hard and needs better tools, or a use case where a new tutorial would be helpful, or documentation for a specific procedure could be added or improved.

**Comment by wjwwood** [comment]
Sorry, I didn't realize this KDL was third party. That changes the whole discussion, in that context this is not isolated to `catkin_make`. In this case, `catkin_make` simply exposed the error. It occurs to me, though I might be wrong, that CMake projects which used the `PKG_CONFIG_PATH` _after_ `find_package`'ing this KDL module will also break.

> However, Catkin imposes numerous constraints on buildscript design which the buildscript writer for an isolated CMake project never seems to have to worry about.

In the issue raised in this thread, KDL setting the PKG_CONFIG_PATH and not restoring it is just wrong, I'm not sure it has anything to do with constraints that catkin puts on package developers.

> This might be a good thing to do for a tightly integrated system like Microsoft Robotics Studio or ERSP, but ROS is not such a system.

I think rosbuild represents a tightly constrained build system which extremely constrains the process in order to get tighter integration. catkin represents an effort to become more conventional while still trying to fill some of the gaps that CMake leaves when dealing with numerous federated software packages, which has lead to a less tightly integrated and coupled system. The pro to that is that it is more conventional, but the downside is that it is more generic and therefore less integrated. Hopefully with the version 2 of the `package.xml` and something like `catkin_simple`, in the near future we can make up some of that integration ground that we have lost.

**Comment by tkruse** [comment]
@jbohren described two things that wasted his time, the PKG_CONFIG_PATH bug
in KDL, but also "Tracking down which dependencies are needed by which
targets across packages...". If we shall focus on specific issues, then
let's hear this one out (either here or as a new issue).

Also you still seem to declare target dependencies across catkin packages
as a feature of catkin, rather than very bad practice that can sadly not be
prevented for the time being.

You stress that 'This means that users can trigger full "rebuilds" and know
that their binaries are all up to date, without worrying about how long it
will take.'.
What good it is to gain 10 minutes every week thanks to a faster build
system, if I lose a full day to find a bug?

On Tue, Oct 22, 2013 at 10:24 PM, Tully Foote notifications@github.comwrote:

> There are users for whom the full power is not as valuable, such as new
> students using ROS. But new students using ROS I do not expect to be
> integrating 3rdparty CMake based libraries which have bugs causing breaking
> side effects.

You do not expect new students in labs who switch from non-ROS software to
mixed ROS/non-ROS software to be integrating their custom/legacy libraries
written in haste with shoddy cmake files? What do you base this expectation
on? Isn't the mixing of ROS with other software (motion planners,
simulators, databases, cognitive architectures) created in the academic
context the every essence of all robotic development?

**Comment by mikepurvis** [comment]
As requested by @tkruse in #543, another tale of woe:

The [workspace in question](https://raw.github.com/kf/install/master/robot.rosinstall) builds fine on my ivy bridge laptop, but fails on the Kingfisher's 32-bit Atom machine. The reported error occurs while generating Lisp messages for kingfisher_msgs—maddeningly, the problem does _not_ occur in a bare workspace containing just that one repo, but it _does_ occur in the original workspace even with everything else having a CATKIN_IGNORE file placed in it.

I still haven't really worked out what's going on, but the rest of the workspace builds without kingfisher_msgs, and kingfisher_msgs builds in a workspace by itself. These are mostly Clearpath packages, so I'm sure it's our fault, but the tooling provides basically zero help in tracking it down—not being able to trust CATKIN_IGNORE is particularly troubling.

(Anyhow, I know this is a ticket and not a tech-support forum; if I have specific questions, I'll post on Askbot. And once I figure this out, I'll update the post with what the issue ended up being...)

---

Blast. Here's the [build output and pkg versions](https://gist.github.com/mikepurvis/7124672) from the broken build, and [here's the change that fixes it](https://github.com/kf/kingfisher/commit/6882ebc2dd60cebd743b9c923936f0e0e0f05e53). Clearly a case of PBKAC—but definitely also a situation where a catkin linter could have warned me that I didn't have the right sequence of incantations for message generation.

I still maintain that the failure mode was bizarre, though; that I could build the kingfisher_msgs package alone, and the build farm even released a deb of it which also worked perfectly fine, and it worked fine on some machines but not others!

**Comment by dirk-thomas** [comment]
I can build the workspace perfectly fine.

Please provide more information. E.g. the full output of your invoked commands. Also information like what system you use. If message generation is related to the issue please also consider posting the version numbers of the related packages you have installed (e.g. genmsg, genlisp).

**Comment by tkruse** [comment]
Some more debugging problems: http://answers.ros.org/question/93691/trying-to-compile-from-source-to-test-fix
Occurs when putting both the mentioned files and geometry_experimental into the same catkin workspace.

**Comment by tkruse** [comment]
Another issue regarding common cmake target namespaces: https://github.com/ros/catkin/issues/477. Not sure whether it ever got fixed.

**Comment by tfoote** [comment]
@tkruse please open separate tickets for separate issues.  

I'm going to close this ticket as **we agree** that we want better debugging tools. In the abstract that is not something achievable without concrete examples of where people are getting tripped up. When you see someone getting tripped up please file a ticket where you think the interaction could be better and we can work to improve the experience for each use case.   Each case will need to be evaluated and some things will be fixable and some not.  For example #477 is purely a limitation of cmake, and we are building on top of cmake and do not have the time and energy to build a replacement. Nor do we want to do so. And it has a very straight forward error saying that it has a conflicting target name and calls out the two files where they collide which makes it pretty clear where the issue is coming from and how to fix it.

**Comment by tkruse** [comment]
It's fair enough if OSRF thinks it's better to spend time and energy on evaluating each case individually and then improving debug tools for problems caused by lack of isolation, rather than just cleanly isolating builds and avoiding those problems in the first place (as cmi does even today). That call is for OSRF to make. 

You can keep this github issue closed, we can still use it to collect all incidents relating to lack of isolation.

**Comment by DuaneNielsen** [comment]
Side effects are bad.  That's one of the reason all the successful programming languages are Context Free.  This was all worked out in 1970's computer science.  In a complex system, it's more important to have tracebility, so you can find and fix errors quickly.  

What's the upside of not having to say exactly what you mean?  You type a few less lines of code?   Weigh that against spending 5 hours working out that installing package X broke package Y.  I know which I would rather have!

**Comment by jbohren** [comment]
@DuaneNielsen Fortunately in the last two years, we've been developing (and using) [catkin_tools](http://github.com/catkin/catkin_tools), which provides parallelized isolated builds, and version [0.4.0](http://github.com/catkin/catkin_tools/milestones/0.4.0%20-%20Second%20Beta%20Announcement) will bring isolated cleaning too!

**Comment by DuaneNielsen** [comment]
nice one, I'll check it out... thanks!
